### PR TITLE
Things

### PR DESCRIPTION
**Pull request by levelcap**

### CODE REVIEW

File: examples/dave_great_copy.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            log_probs = F.softmax(logits, dim=-1)
+            if sample:
+                prev = torch.multinomial(log_probs, num_samples=1)
+            else:
+                _, prev = torch.topk(log_probs, k=1, dim=-1)
+            output = torch.cat((output, prev), dim=1)
+    return output
+
+def run_model():
+    parser = argparse.ArgumentParser()
```
Comment (gcastanon) [inline examples/dave_great_copy.py:56]: THIS PARSES ARGUMENTS!  ...I bet you knew this.

File: examples/dave_great_copy.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    parser.add_argument("--temperature", type=float, default=1.0)
+    parser.add_argument("--top_k", type=int, default=0)
+    parser.add_argument('--unconditional', action='store_true', help='If true, unconditional generation.')
+    args = parser.parse_args()
+    print(args)
+
+    if args.batch_size == -1:
+        args.batch_size = 1
+    assert args.nsamples % args.batch_size == 0
+
```
Comment (gcastanon) [inline examples/dave_great_copy.py:71]: It's important to set everybody's random seed to a fixed number so that you don't different output for the same input when you run multiple times.  That's all this is doing.

File: examples/dave_great_copy.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    np.random.seed(args.seed)
+    torch.random.manual_seed(args.seed)
+    torch.cuda.manual_seed(args.seed)
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (gcastanon) [inline examples/dave_great_copy.py:75]: Notably this is the line where we tell torch if it gets to use a GPU or has to use a CPU.  You can use a CPU for now, but should eventually learn how to set up a GPU.

File: examples/dave_great_copy.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    enc = GPT2Tokenizer.from_pretrained(args.model_name_or_path)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:77]: A "Tokenizer" takes in words/sentences and transforms them into vectors (e.g. [0,.5,3,.75,82,0,]) that we can stuff into a neural network.  I do not know what tokenizer these guys are using, but it's pretrained, so just use it.  =)

File: examples/dave_great_copy.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    model = GPT2LMHeadModel.from_pretrained(args.model_name_or_path)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:78]: This loads the predictive model.  aka the actual neural network that the output of the tokenizer is going to go into. 
 That's about it.

File: examples/dave_great_copy.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    model.to(device)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:79]: Send the model to the device

File: examples/dave_great_copy.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    model.eval()
```
Comment (gcastanon) [inline examples/dave_great_copy.py:80]: Tell your model that it's in evaluation mode, not training mode: don't change the model in response to data, just run data through.

File: examples/dave_great_copy.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+    if args.length == -1:
+        args.length = model.config.n_ctx // 2
+    elif args.length > model.config.n_ctx:
+        raise ValueError("Can't get samples longer than window size: %s" % model.config.n_ctx)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:85]: n_ctx means "Length of Context Window".  aka "How far backwards and forwards am I paying attention, as counted in words".  So if n_ctx=5, for my previous sentence, I would have "attention, as counted in words".  If n_ctx=2, "in words"

File: examples/dave_great_copy.py (hunk 4 above)
Comment (levelcap) [reply examples/dave_great_copy.py:77]: Does that mean that the `--model_name_or_path` argument is what we'd use to pass in the actual prompts?

nevermind! I should wait until the end of the commentary.

File: examples/dave_great_copy.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    while True:
+        context_tokens = []
+        if not args.unconditional:
+            raw_text = input("Model prompt >>> ")
```
Comment (gcastanon) [inline examples/dave_great_copy.py:90]: Give a prompt.

File: examples/dave_great_copy.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+            while not raw_text:
+                print('Prompt should not be empty!')
+                raw_text = input("Model prompt >>> ")
+            context_tokens = enc.encode(raw_text)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:94]: Turn your text into vectors via the tokenizer.

File: examples/dave_great_copy.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+            generated = 0
+            for _ in range(args.nsamples // args.batch_size):
+                out = sample_sequence(
+                    model=model, length=args.length,
+                    context=context_tokens,
+                    start_token=None,
+                    batch_size=args.batch_size,
+                    temperature=args.temperature, top_k=args.top_k, device=device
+                )
+                out = out[:, len(context_tokens):].tolist()
```
Comment (gcastanon) [inline examples/dave_great_copy.py:104]: Feed your stuff through the model some times to generate some output vectors

File: examples/dave_great_copy.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                for i in range(args.batch_size):
+                    generated += 1
+                    text = enc.decode(out[i])
```
Comment (gcastanon) [inline examples/dave_great_copy.py:107]: Turn those vectors back into words

File: examples/dave_great_copy.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+                    print("=" * 40 + " SAMPLE " + str(generated) + " " + "=" * 40)
+                    print(text)
+            print("=" * 80)
+        else:
```
Comment (gcastanon) [inline examples/dave_great_copy.py:111]: This is just for the case where you declined to provide it with a cue.  This is the "JUST SAY SOMETHING, AI.  ANYTHING!"

File: examples/dave_great_copy.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    """
+    if k == 0:
+        return logits
+    else:
+        values = torch.topk(logits, k)[0]
+        batch_mins = values[:, -1].view(-1, 1).expand_as(logits)
+        return torch.where(logits < batch_mins, torch.ones_like(logits) * -1e10, logits)
+
+def sample_sequence(model, length, start_token=None, batch_size=None, context=None, temperature=1, top_k=0, device='cuda', sample=True):
+    if start_token is None:
```
Comment (gcastanon) [inline examples/dave_great_copy.py:33]: This is just deciding if you gave it prompt/cue or not.

File: examples/dave_great_copy.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    output = context
+    past = None
+    with torch.no_grad():
+        for i in trange(length):
+            logits, past = model(prev, past=past)
+            logits = logits[:, -1, :] / temperature
+            logits = top_k_logits(logits, k=top_k)
+            log_probs = F.softmax(logits, dim=-1)
+            if sample:
+                prev = torch.multinomial(log_probs, num_samples=1)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:49]: This option says: Randomly sample remaining words based on their probability. So I'd get "poop" 2 times in 3 and "shit" the other time.

File: examples/dave_great_copy.py [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+            else:
+                _, prev = torch.topk(log_probs, k=1, dim=-1)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:51]: This option says: Just take the most probable word.

File: examples/dave_great_copy.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+        context = torch.full((batch_size, 1), start_token, device=device, dtype=torch.long)
+    prev = context
+    output = context
+    past = None
+    with torch.no_grad():
+        for i in trange(length):
+            logits, past = model(prev, past=past)
+            logits = logits[:, -1, :] / temperature
+            logits = top_k_logits(logits, k=top_k)
+            log_probs = F.softmax(logits, dim=-1)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:47]: "...and now weight heavily towards the guy with the biggest score"

File: examples/dave_great_copy.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+        assert context is None, 'Specify exactly one of start_token and context!'
+        context = torch.full((batch_size, 1), start_token, device=device, dtype=torch.long)
+    prev = context
+    output = context
+    past = None
+    with torch.no_grad():
+        for i in trange(length):
+            logits, past = model(prev, past=past)
+            logits = logits[:, -1, :] / temperature
+            logits = top_k_logits(logits, k=top_k)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:46]: This says "only pay attention to the top K probabilities"

File: examples/dave_great_copy.py [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+        context = torch.tensor(context, device=device, dtype=torch.long).unsqueeze(0).repeat(batch_size, 1)
+    else:
+        assert context is None, 'Specify exactly one of start_token and context!'
+        context = torch.full((batch_size, 1), start_token, device=device, dtype=torch.long)
+    prev = context
+    output = context
+    past = None
+    with torch.no_grad():
+        for i in trange(length):
+            logits, past = model(prev, past=past)
```
Comment (gcastanon) [inline examples/dave_great_copy.py:44]: logits, here, is a probably distribution over a bunch of words.  so like, we say "poop" with 50% probability, "shit" with 25% probability, and something else with 25% probability.

File: examples/dave_great_copy.py [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+            logits = logits[:, -1, :] / temperature
```
Comment (gcastanon) [inline examples/dave_great_copy.py:45]: This is normalization.  Its purpose is to prevent the neural network from gravitating towards using the same words all the time.
